Cargo.toml

```diff
@@ -17,14 +20,15 @@ rpassword = "4.0"
 secstr = "0.3"
 serde = { version = "1.0", features = ["derive"] }
 serde_cbor = "0.10"
+thrussh-keys = { version = "0.20", optional = true }
+tokio = { version = "1.0", features = ["net"], optional = true }
```

Comment:
It is possible that I’m misunderstanding something about cargo’s dependency solver, the docs are wrong, or the cargo defaults are just the worst possible choice.

Afaiu, the above is equivalent to “^1.0”, which supposedly means that cargo can use a newer version if it starts with “1.0”, eg 1.0.3, 1.0.10, etc

This is mucho semantico, because the dependency is supposed to only increment the zero if it would somehow break consumers which expect the previous API. Right?


Now, what I see a lot is that crates actually depend on a three-digit version, eg `quinn` depends on `tokio ^1.0.1`. What I would expect to happen if I set this to `1.4` here, and then use this crate from librad is that I’ll end up with two versions of tokio (which is a situation also referred to as “async hell”).

Is there some semantic wishywashy I’m missing? Or is it just really bad practice to have such strict version bounds?

Reply:
Yeah, so what I was missing is the difference between versions that start with zero, and ones that don’t (zeroes are special).

```
^1.2.3  :=  >=1.2.3, <2.0.0
^1.2    :=  >=1.2.0, <2.0.0
^1      :=  >=1.0.0, <2.0.0
^0.2.3  :=  >=0.2.3, <0.3.0
^0.2    :=  >=0.2.0, <0.3.0
^0.0.3  :=  >=0.0.3, <0.0.4
^0.0    :=  >=0.0.0, <0.1.0
^0      :=  >=0.0.0, <1.0.0
```

**tl;dr** Since tokio is > 1.0, it doesn’t really matter. But using `1.4` here will auto-upgrade all consumers.

Reply:
I suppose a superior software versioning system would be to use the hash of the AST on module level, and then specify in your dependencies file how many **leading zeroes** is shall have.